File: frontend/src/planning/surfaceBreakdown.ts
import type { RouteVisualization, SurfaceInsight } from "../api";

export interface SurfaceCategoryStat {
  riderCategory: string;
  label: string;
  distanceKm: number;
  percentage: number;
  color: string;
}

const CATEGORY_COLORS: Record<string, string> = {
  Road: "#2563eb",
  Gravel: "#854d0e",
  Trail: "#16a34a",
  Unknown: "#ef4444",
};

const CATEGORY_ORDER = ["Road", "Gravel", "Trail", "Unknown"];

export function buildSurfaceCategoryBreakdown(
  route: RouteVisualization,
  totalKm: number,
): SurfaceCategoryStat[] {
  const totals = new Map<string, number>();

  for (const segment of route.surface_segments) {
    const key = segment.rider_category ?? segment.surface;
    totals.set(key, (totals.get(key) ?? 0) + Math.max(0, segment.end_km - segment.start_km));
  }

  return CATEGORY_ORDER.map((category) => {
    const distanceKm = totals.get(category) ?? 0;
    return {
      riderCategory: category,
      label: category,
      distanceKm: Math.round(distanceKm * 10) / 10,
      percentage: totalKm > 0 ? Math.round((distanceKm / totalKm) * 1000) / 10 : 0,
      color: CATEGORY_COLORS[category] ?? "#9ca3af",
    };
  }).filter((row) => row.distanceKm > 0 || row.riderCategory === "Unknown");
}

export function surfaceSegmentMatchesSelection(
  segment: { rider_category?: string; surface: string; osm_surface: string | null },
  selected: string | null,
): boolean {
  if (!selected) {
    return true;
  }
  return (
    segment.rider_category === selected ||
    segment.surface === selected ||
    (segment.osm_surface ?? "unknown") === selected
  );
}

export function formatSurfaceKmRange(startKm: number, endKm: number): string {
  return `KM ${Math.round(startKm)} → ${Math.round(endKm)}`;
}

export function surfaceInsightDetail(insight: SurfaceInsight): string {
  const range = formatSurfaceKmRange(insight.start_km, insight.end_km);
  if (insight.subcategory) {
    return `${range} · ${insight.subcategory}`;
  }
  return range;
}

export function surfaceSourceLabel(source: string): string {
  switch (source) {
    case "osm_tag":
      return "OSM surface tag";
    case "highway_inferred":
      return "Inferred from road type";
    case "propagated":
      return "Propagated from neighbours";
    case "insufficient_evidence":
      return "Insufficient evidence";
    case "unmatched":
      return "No OSM match";
    default:
      return source;
  }
}

File: src/surface_breakdown.py
"""Aggregate rider-facing surface categories for the Surface page."""

from dataclasses import dataclass

from route_visualization import RouteVisualization
from surface_types import RIDER_CATEGORY_COLORS, RiderCategory


@dataclass(frozen=True)
class SurfaceCategoryStat:
    rider_category: str
    label: str
    distance_km: float
    percentage: float
    color: str


@dataclass(frozen=True)
class SurfaceSubcategoryStat:
    rider_category: str
    rider_subcategory: str
    distance_km: float
    percentage: float


def build_surface_category_breakdown(
    visualization: RouteVisualization,
    total_km: float,
) -> list[SurfaceCategoryStat]:
    totals: dict[str, float] = {}

    for segment in visualization.surface_segments:
        totals[segment.rider_category] = totals.get(segment.rider_category, 0.0) + max(
            0.0,
            segment.end_km - segment.start_km,
        )

    order = [RiderCategory.ROAD, RiderCategory.GRAVEL, RiderCategory.TRAIL, RiderCategory.UNKNOWN]
    stats: list[SurfaceCategoryStat] = []
    for category in order:
        distance_km = totals.get(category.value, 0.0)
        stats.append(
            SurfaceCategoryStat(
                rider_category=category.value,
                label=category.value,
                distance_km=round(distance_km, 1),
                percentage=round((distance_km / total_km * 100) if total_km > 0 else 0.0, 1),
                color=RIDER_CATEGORY_COLORS[category],
            )
        )
    return stats


def build_surface_subcategory_breakdown(
    visualization: RouteVisualization,
    total_km: float,
) -> list[SurfaceSubcategoryStat]:
    totals: dict[tuple[str, str], float] = {}

    for segment in visualization.surface_segments:
        key = (segment.rider_category, segment.rider_subcategory)
        totals[key] = totals.get(key, 0.0) + max(0.0, segment.end_km - segment.start_km)

    stats: list[SurfaceSubcategoryStat] = []
    for (category, subcategory), distance_km in sorted(totals.items(), key=lambda item: item[1], reverse=True):
        stats.append(
            SurfaceSubcategoryStat(
                rider_category=category,
                rider_subcategory=subcategory,
                distance_km=round(distance_km, 1),
                percentage=round((distance_km / total_km * 100) if total_km > 0 else 0.0, 1),
            )
        )
    return stats
